CASSGO-38 Marshal error messages enhancement

tengu-alt opened the pull request:
This PR provides marshal/unmarshal error messages enhancement.
According to #671 errors were updated by the supported types list.

worryg0d commented on marshal.go:
```diff
@@ -333,7 +333,7 @@ func marshalVarchar(info TypeInfo, value interface{}) ([]byte, error) {
 		return rv.Bytes(), nil
 	}
-	return nil, marshalErrorf("can not marshal %T into %s", value, info)
+	return nil, marshalErrorf("can not marshal %T into %s. Accepted types: Marshaler, string, []byte, unsetColumn.", value, info)
```
`unsetColumn` is not an exported type, so the driver doesn't have to expose it to the user. There is a `UnsetValue` global var of type `unsetColumn` which is meant to be used to ignore writing. You can find it [here](https://github.com/apache/cassandra-gocql-driver/blob/2a5056d6b1d1d479ca5ee97ddd1d95a9d8e8c4e2/frame.go#L50).

tengu-alt replied:
Thank you, refactored

worryg0d replied:
I meant to change unsetColumn to UnsetValue, not to delete it at all

tengu-alt replied:
I see, fixed.

worryg0d commented on marshal.go:
```diff
@@ -2280,7 +2280,7 @@ func marshalUDT(info TypeInfo, value interface{}) ([]byte, error) {
 
 	if k.Kind() != reflect.Struct || !k.IsValid() {
-		return nil, marshalErrorf("cannot marshal %T into %s", value, info)
+		return nil, marshalErrorf("cannot marshal %T into %s. Accepted types: Marshaler, unsetColumn, UDTMarshaler, UDTMarshaler, map[string]interface{}, struct.", value, info)
```
Here **UDTMarshaler** is duplicated

tengu-alt replied:
fixed.

ribaraka commented:
https://issues.apache.org/jira/browse/CASSGO-38

joao-r-reis commented on marshal.go:
```diff
@@ -2208,7 +2210,7 @@ func unmarshalTuple(info TypeInfo, data []byte, value interface{}) error {
 	}
 
-	return unmarshalErrorf("cannot unmarshal %s into %T", info, value)
+	return unmarshalErrorf("cannot unmarshal %s into %T. Accepted types: struct, []interface{}, array, slice.", info, value)
```
Unmarshaler is also accepted I think. Also the accepted types other than `[]interface{}` have to be a pointer if I'm reading the code correctly (so `*struct, *array, *slice`)

tengu-alt replied:
Good catch! Thank you.

joao-r-reis commented on marshal.go:
```diff
@@ -2280,7 +2282,7 @@ func marshalUDT(info TypeInfo, value interface{}) ([]byte, error) {
 
 	if k.Kind() != reflect.Struct || !k.IsValid() {
-		return nil, marshalErrorf("cannot marshal %T into %s", value, info)
+		return nil, marshalErrorf("cannot marshal %T into %s. Accepted types: Marshaler, UDTMarshaler, map[string]interface{}, struct, UnsetValue.", value, info)
```
It looks like `struct` is also accepted here

tengu-alt replied:
Explain please, because `struct` is already here.

joao-r-reis replied:
Oh I'm sorry, I was probably looking at the wrong place

joao-r-reis commented on marshal.go:
```diff
@@ -1706,7 +1708,7 @@ func unmarshalList(info TypeInfo, data []byte, value interface{}) error {
 		return nil
 	}
-	return unmarshalErrorf("can not unmarshal %s into %T", info, value)
+	return unmarshalErrorf("can not unmarshal %s into %T. Accepted types: slice, array.", info, value)
```
I think it has to be `*slice, *array`

tengu-alt replied:
Fixed.

joao-r-reis commented on marshal.go:
```diff
@@ -2392,7 +2394,7 @@ func unmarshalUDT(info TypeInfo, data []byte, value interface{}) error {
 	k := rv.Elem()
 	if k.Kind() != reflect.Struct || !k.IsValid() {
-		return unmarshalErrorf("cannot unmarshal %s into %T", info, value)
+		return unmarshalErrorf("cannot unmarshal %s into %T. Accepted types: Unmarshaler, UDTUnmarshaler, *map[string]interface{}, struct.", info, value)
```
`*struct`

tengu-alt replied:
Fixed.

tengu-alt committed:
Improve error messages for marshalling

Enhancement for marshal/unmarshal error messages.
Errors were updated by the supported types list.

patch by Oleksandr Luzhniy; reviewed by João Reis, James Hartig, Bohdan Siryk, for CASSGO-38

jameshartig commented:
It's not very common in Go for the error to explicitly provide the supported types but I can see how that's useful. Ideally a developer would be able to know the supported types before encountering an error. In #671 what made you think that timestamp would support an int? Was it just common sense or did you read something indicating as much?

I'd like to see improvements to the docs and a general improvement in the supported types. That said, nothing is wrong with this PR.

tengu-alt commented:
> In https://github.com/apache/cassandra-gocql-driver/issues/671 what made you think that timestamp would support an int? Was it just common sense or did you read something indicating as much?

I don't clearly understand your question. The driver already supports `Timestamp` marshal into `int64`, as said in the [codebase](https://github.com/apache/cassandra-gocql-driver/blob/trunk/marshal.go#L1273) and [documentation](https://github.com/apache/cassandra-gocql-driver/blob/trunk/marshal.go#L87).

jameshartig commented:
> I don't clearly understand your question.

In #671 you ran into:

> 2016/03/08 08:35:36 can not marshal int into timestamp

Which meant you sent an int, that should be supported. I'm trying to understand how we can improve the documentation or API to make it more obvious which types are supported that doesn't involve someone having to parse an error after already writing code which assumed a certain type.

tengu-alt commented:
> In https://github.com/apache/cassandra-gocql-driver/issues/671 you ran into:

That was not me who ran into the error, the #671 was created by another person.

As Cassandra [doc](https://docs.datastax.com/en/cql-oss/3.x/cql/cql_reference/timestamp_type_r.html) says: `timestamp` is encoded as `int64`. The default `int` in GO is a 32bit integer, so it is not supported.

IMHO the documentation in the `marshal.go` is pretty obvious, and it appears when you call the `func Marshal()`.

jameshartig commented:
> That was not me who ran into the error, the https://github.com/apache/cassandra-gocql-driver/issues/671 was created by another person.

Oh apologies for the confusion.

> The default int in GO is a 32bit integer, so it is not supported.

Only on 32-bit systems, which are increasingly rare. We should be fine marshalling into Timestamp but maybe we just couldn't safely unmarshal

tengu-alt commented:
@joao-r-reis I have updated the `Changelog.md`, should it be merged?
